encoding/protobuf/v2/error.go

```diff
@@ -0,0 +1,247 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package v2
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes/any"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/internal/grpcerrorcodes"
+	"go.uber.org/yarpc/yarpcerrors"
+	rpc_status "google.golang.org/genproto/googleapis/rpc/status"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/anypb"
+)
+
+const (
+	// format for converting error details to string
+	_errDetailsFmt = "[]{ %s }"
+	// format for converting a single message to string
+	_errDetailFmt = "%s{%s}"
+)
+
+var _ error = (*pberror)(nil)
+
+type pberror struct {
+	code    yarpcerrors.Code
+	message string
+	details []*any.Any
+}
+
+func (err *pberror) Error() string {
+	var b strings.Builder
+	b.WriteString("code:")
+	b.WriteString(err.code.String())
+	if err.message != "" {
+		b.WriteString(" message:")
+		b.WriteString(err.message)
+	}
+	return b.String()
+}
+
+// NewError returns a new YARPC protobuf error. To access the error's fields,
+// use the yarpcerrors package APIs for the code and message, and the
+// `GetErrorDetails(error)` function for error details. The `yarpcerrors.Details()`
+// will not work on this error.
+//
+// If the Code is CodeOK, this will return nil.
+func NewError(code yarpcerrors.Code, message string, options ...ErrorOption) error {
+	if code == yarpcerrors.CodeOK {
+		return nil
+	}
+	pbErr := &pberror{
+		code:    code,
+		message: message,
+	}
+	for _, opt := range options {
+		if err := opt.apply(pbErr); err != nil {
+			return err
+		}
+	}
+	return pbErr
+}
+
+// GetErrorDetails returns the error details of the error.
+//
+// This method supports extracting details from wrapped errors.
+//
+// Each element in the returned slice of interface{} is either a proto.Message
+// or an error to explain why the element is not a proto.Message, most likely
+// because the error detail could not be unmarshaled.
+func GetErrorDetails(err error) []interface{} {
+	var target *pberror
```

Comment:
can we validate if err is nil and return it from here instead of continuing?

Reply:
error.As internally checks for whether the first argument(err) is nil or not. see: https://cs.opensource.google/go/go/+/refs/tags/go1.17.6:src/errors/wrap.go;l=90